linalg/sundials.hpp

```diff
@@ -23,384 +23,433 @@
+   /// Construct a parallel wrapper to SUNDIALS' ARKode integrator.
+   /** @param[in] comm The MPI communicator used to partition the ODE system.
+       @param[in] type Specifies the RK method type:
+                       - EXPLICIT - explicit RK method (default)
+                       - IMPLICIT - implicit RK method
+                       - IMEX     - implicit-explicit ARK method */
+   ARKStepSolver(MPI_Comm comm, Type type = EXPLICIT);
 #endif
 
-   /// Specify the scalar relative and scalar absolute tolerances.
-   void SetSStolerances(double reltol, double abstol);
+   /** @brief Initialize ARKode: calls ARKStepCreate() to create the ARKStep
+       memory and set some defaults.
+
+       If the ARKStep has already been created, it checks if the problem size
+       has changed since the last call to Init(). If the problem is the same
+       then ARKStepReInit() will be called in the next call to Step(). If the
+       problem size has changed, the ARKStep memory is freed and realloced
+       for the new problem size. */
+   /** @param[in] f_ The TimeDependentOperator that defines the ODE system
+
+       @note All other methods must be called after Init(). */
+   void Init(TimeDependentOperator &f_);
+
+   /// Integrate the ODE with ARKode using the specified step mode.
+   /**
+       @param[in,out] x  On output, the solution vector at the requested output
+                         time, tout = @a t + @a dt
+       @param[in,out] t  On output, the output time reached
+       @param[in,out] dt On output, the last time step taken
+
+       @note On input, the values of @a t and @a dt are used to compute desired
+       output time for the integration, tout = @a t + @a dt.
+   */
+   virtual void Step(Vector &x, double &t, double &dt);
+
+   /// Attach an MFEM linear solver solver to ARKode.
+   void UseMFEMLinearSolver();
 
-   /// Set a custom Jacobian system solver for implicit methods.
-   void SetLinearSolver(SundialsODELinearSolver &ls_spec);
+   /// Attach a SUNDIALS linear solver to ARKode.
+   void UseSundialsLinearSolver();
 
-   /** @brief ARKode supports two modes, specified by itask: ARK_NORMAL
-       (default) and ARK_ONE_STEP. */
-   /** In the ARK_NORMAL mode, the solver steps until it reaches or passes
-       tout = t + dt, where t and dt are specified in Step(), and then
-       interpolates to obtain y(tout). In the ARK_ONE_STEP mode, it takes one
-       internal step and returns. */
+   /// Attach an MFEM mass matrix linear solver to ARKode.
+   /** @param[in] tdep    An integer flag indicating if the mass matrix is time
+                          dependent (1) or time independent (0) */
+   void UseMFEMMassLinearSolver(int tdep);
+
+   /// Attach a SUNDIALS mass matrix linear solver to ARKode.
+   /** @param[in] tdep    An integer flag indicating if the mass matrix is time
+                          dependent (1) or time independent (0) */
+   void UseSundialsMassLinearSolver(int tdep);
```

Comment:
Similar to above, which `MFEM`/`SUNDIALS` solvers will be used when `UseMFEMMassLinearSolver`/`UseSundialsMassLinearSolver` are called?